File: sample-input/benchmarks/homogeneous-one-group/homogeneous-one-group.py
import numpy
from openmoc import *
import openmoc.log as log
import openmoc.plotter as plotter
from openmoc.options import Options


###############################################################################
#######################   Main Simulation Parameters   ########################
###############################################################################

options = Options()

num_threads = options.getNumThreads()
track_spacing = options.getTrackSpacing()
num_azim = options.getNumAzimAngles()
tolerance = options.getTolerance()
max_iters = options.getMaxIterations()

log.set_log_level('NORMAL')

log.py_printf('TITLE', 'Simulating a one group homogeneous infinite medium...')
log.py_printf('HEADER', 'The reference keff = 1.43...')


###############################################################################
#######################   Main Simulation Parameters   ########################
###############################################################################

log.py_printf('NORMAL', 'Creating materials...')

infinite_medium = Material(name='1-group infinite medium')
infinite_medium.setNumEnergyGroups(1)
infinite_medium.setSigmaA(numpy.array([0.069389522]))
infinite_medium.setSigmaF(numpy.array([0.0414198575]))
infinite_medium.setNuSigmaF(numpy.array([0.0994076580]))
infinite_medium.setSigmaS(numpy.array([0.383259177]))
infinite_medium.setChi(numpy.array([1.0]))
infinite_medium.setSigmaT(numpy.array([0.452648699]))


###############################################################################
###########################   Creating Surfaces   #############################
###############################################################################

log.py_printf('NORMAL', 'Creating surfaces...')

left = XPlane(x=-100.0, name='left')
right = XPlane(x=100.0, name='right')
top = YPlane(y=100.0, name='top')
bottom = YPlane(y=-100.0, name='bottom')

left.setBoundaryType(REFLECTIVE)
right.setBoundaryType(REFLECTIVE)
top.setBoundaryType(REFLECTIVE)
bottom.setBoundaryType(REFLECTIVE)


###############################################################################
#############################   Creating Cells   ##############################
###############################################################################

log.py_printf('NORMAL', 'Creating cells...')

cell = CellBasic()
cell.setMaterial(infinite_medium)
cell.addSurface(halfspace=+1, surface=left)
cell.addSurface(halfspace=-1, surface=right)
cell.addSurface(halfspace=+1, surface=bottom)
cell.addSurface(halfspace=-1, surface=top)


###############################################################################
#                            Creating Universes
###############################################################################

log.py_printf('NORMAL', 'Creating universes...')

root_universe = Universe(name='root universe')
root_universe.addCell(cell)


###############################################################################
##########################   Creating the Geometry   ##########################
###############################################################################

log.py_printf('NORMAL', 'Creating geometry...')

geometry = Geometry()
geometry.setRootUniverse(root_universe)
geometry.initializeFlatSourceRegions()


###############################################################################
########################   Creating the TrackGenerator   ######################
###############################################################################

log.py_printf('NORMAL', 'Initializing the track generator...')

track_generator = TrackGenerator(geometry, num_azim, track_spacing)
track_generator.setNumThreads(num_threads)
track_generator.generateTracks()


###############################################################################
###########################   Running a Simulation   ##########################
###############################################################################

solver = CPUSolver(geometry, track_generator)
solver.setNumThreads(num_threads)
solver.setSourceConvergenceThreshold(tolerance)
solver.convergeSource(max_iters)
solver.printTimerReport()

log.py_printf('TITLE', 'Finished')

File: profile/models/homogeneous/homogeneous-one-group.cpp
#include "../../../src/CPUSolver.h"
#include "../../../src/log.h"
#include <array>
#include <iostream>

int main() {

  /* Define simulation parameters */
  #ifdef OPENMP
  int num_threads = omp_get_num_procs();
  #else
  int num_threads = 1;
  #endif
  double azim_spacing = 0.1;
  int num_azim = 4;
  double tolerance = 1e-5;
  int max_iters = 1000;

  /* Set logging information */
  set_log_level("NORMAL");
  log_printf(TITLE, "Simulating a one group homogeneous infinite medium...");
  log_printf(HEADER, "The reference keff = 1.43...");

  /* Create materials */
  log_printf(NORMAL, "Creating materials...");
  Material* infinite_medium = new Material();
  infinite_medium->setNumEnergyGroups(1);
  double sigmaF[1] = {0.0414198575};
  double nuSigmaF[1] = {0.0994076580};
  double sigmaS[1] = {0.383259177};
  double chi[1] = {1.0};
  double sigmaT[1] = {0.452648699};
  infinite_medium->setSigmaF(sigmaF, 1);
  infinite_medium->setNuSigmaF(nuSigmaF, 1);
  infinite_medium->setSigmaS(sigmaS, 1);
  infinite_medium->setChi(chi, 1);
  infinite_medium->setSigmaT(sigmaT, 1);

  /* Create surfaces */
  log_printf(NORMAL, "Creating surfaces...");
  double L = 200.0;
  XPlane left(-L/2);
  XPlane right(L/2);
  YPlane top(L/2);
  YPlane bottom(-L/2);

  left.setBoundaryType(REFLECTIVE);
  right.setBoundaryType(REFLECTIVE);
  top.setBoundaryType(REFLECTIVE);
  bottom.setBoundaryType(REFLECTIVE);

  /* Create cells */
  log_printf(NORMAL, "Creating cells...");

  Cell* cell = new Cell();
  cell->setFill(infinite_medium);
  cell->addSurface(+1, &left);
  cell->addSurface(-1, &right);
  cell->addSurface(+1, &bottom);
  cell->addSurface(-1, &top);

  /* Create universes */
  log_printf(NORMAL, "Creating universes...");

  Universe* root_universe = new Universe();
  root_universe->addCell(cell);

  /* Create the geometry */
  log_printf(NORMAL, "Creating geometry...");
  Geometry geometry;
  geometry.setRootUniverse(root_universe);

  /* Generate tracks */
  log_printf(NORMAL, "Initializing the track generator...");
  TrackGenerator track_generator(&geometry, num_azim, azim_spacing);
  track_generator.setNumThreads(num_threads);
  track_generator.generateTracks();

  /* Run simulation */
  CPUSolver solver(&track_generator);
  solver.setNumThreads(num_threads);
  solver.setConvergenceThreshold(tolerance);
  solver.computeEigenvalue(max_iters);
  solver.printTimerReport();

  return 0;
}
